### PR TITLE
DOC: sparse: migration to sparray guide updates

### PR DESCRIPTION
**Pull request by dschult**
These updates to the migration to sparray guide primarily serve to give the reader tools to ease the migration process. 

The biggest change is some code that monkeypatches `__mul__` and `__pow__` to raise exceptions when sparse matrices use`*` and `**` for matrix multiplication/power. This tool helps identify expressions that need @ instead of *.

There are some wording changes to describe how to handle:
- cases where library calls require dtype `int32` index arrays but sparray creates int64.
- random_state parameters for the new function `random_array`.

### CODE REVIEW

**Comment by dschult** [comment]
The CI failure on docs seems to be a Userwarning from matplotlib on a doc_string example from `interpolate.interpnd.CloughTocher2DInterpolater`. The example uses `plot.legend()` and the warning says:
```
UserWarning: Creating legend with loc="best" can be slow with large amounts of data.
```
The default value for `loc` is "best". so to remove the warning we need to use `plot.legend(loc='upper right')` (that leaves the picture the same without making matplotlib compute where to put it.)  But its strange to need that here. Right?
I can make a commit with this "fix" in it. Should I?

I've checked the resulting docs and they look fine to me.  So I'm saying this is fine despite the red X.

File: doc/source/reference/sparse.migration_to_sparray.rst [hunk 1]
```diff
@@ -245,17 +314,160 @@ Removed methods and attributes
...
+(see `gh-18509 <https://github.com/scipy/scipy/pull/18509>`__ for more details).
+This means you may get int64 indexing when you used to get int32.
+You can control this by setting the ``dtype`` before instantiating, or by
+recasting after construction.
+
+Two sparse utility functions can help with handling the index dtype.
+Use ``get_index_dtype(arrays, maxval, check_contents)`` while creating indices
+to find an appropriate dtype (int32 or int64) to use for your compressed indices.
+
+Use ``scipy.sparse.safely_cast_index_arrays`` for recasting after construction,
```
Comment (perimosocordiae) [inline doc/source/reference/sparse.migration_to_sparray.rst:441]: As of gh-21858, this isn't yet exposed publicly, right?

**Review by perimosocordiae — APPROVED** [review]

File: doc/source/reference/sparse.migration_to_sparray.rst (hunk 1 above)
Comment (dschult) [reply doc/source/reference/sparse.migration_to_sparray.rst]: Good catch!
I edited this line to remove the path to the function (just like the description of `get_index_dtype` above).
And then I edited the code section below to include the path for both of these helper functions.

Thanks!!

**Comment by perimosocordiae** [comment]
Merged!

**Pull request merged**